File: AlgorithmDataStructure/LeetCode/98-ValidBST/valid.swift

import Foundation

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}


class Solution {
    
    
    //中序遍历
    func isValidBST(_ root: TreeNode?) -> Bool {
        var stack = [TreeNode]()
        var inorder = Int.min
        
        var tempRoot = root
        
        while stack.isEmpty == false || tempRoot != nil {
            while tempRoot != nil {
                stack.append(tempRoot!)
                tempRoot = tempRoot!.left
            }
            
            tempRoot = stack.removeLast()
            if tempRoot!.val <= inorder { return false }
            inorder = tempRoot!.val
            tempRoot = tempRoot!.right
        }
        return true
    }
    
    
}

let n11 = TreeNode(1)
let n12 = TreeNode(3)
var n1 = TreeNode(2)
n1.left = n11
n1.right = n12


let s = Solution()
print(s.isValidBST(n1))

File: AlgorithmDataStructure/LeetCode/98-ValidBST/valid.c

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

bool helper(struct TreeNode* root, long lower, long upper){
    if (root == NULL) { return true; }
    if (root->val <= lower || root->val >= upper) { return false; }
    return helper(root->left, lower, root->val) && helper(root->right, root->val, upper);
}


bool isValidBST(struct TreeNode* root){
    return helper(root, LONG_MIN, LONG_MAX);
}


int main(int argc, const char * argv[]) {
    
    struct TreeNode n11 = {1, NULL, NULL};
    struct TreeNode n12 = {3, NULL, NULL};
    struct TreeNode n1 = {2, &n11, &n12};
    
    struct TreeNode n21 = {7, NULL, NULL};
    struct TreeNode n22 = {9, NULL, NULL};
    struct TreeNode n2 = {8, &n21, &n22};

    struct TreeNode n = {7, &n1, &n2};

    printf("= %d\n", isValidBST( &n));
    
    /**
     [] = true
     [1] = true
     [2,1] = true
     [1,2] = false
     
     [4,1,6,null,null,2,7] = false  右侧 > ，2<4
     [4,1,6,null,null,5,7] = true
     
     二叉搜索树，左子树所有值都要小于根节点，右子树的所有值都要大于根节点
     
     */
    
    return 0;
}

















